File: Strata/trunk/java/components/StrataCommon/src/strata1/common/producerconsumer/IConsumer.java
// ##########################################################################
// # File Name:	IConsumer.java
// #
// # Copyright:	2012, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataCommon Framework.
// #
// #   			The StrataCommon Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataCommon Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataCommon
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata1.common.producerconsumer;

/****************************************************************************
 * 
 * @author 		
 *     Sapientia Systems
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public 
interface IConsumer<
    T,
    C extends IConsumer<T,C,R,S>,
    R extends IRouter<T,C,R,S>,
    S extends ISelector<T>>
{
    public void
    setSource(R source);
    
    public void
    setSelector(S selector);
    
    public void
    clearSource();
    
    public void
    putElement(T element);
    
    public R
    getSource();
    
    public S
    getSelector();
    
    public int
    getWaitingCount();
    
    public void
    startConsuming();
    
    public void
    stopConsuming();
    
    public boolean
    isConsuming();
}

// ##########################################################################

File: legacy/csharp/components/StrataCommon/ProducerConsumer/IConsumer.cs
//  ##########################################################################
//  # File Name: IConsumer.cs
//  # Copyright: 2016, Sapientia Systems, LLC.
//  ##########################################################################

using System;

namespace Strata.Common.ProducerConsumer
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// 
    /// </summary>
    ///  
    public
    interface IConsumer<T>
    {
        IConsumer<T>
        SetSelector(ISelector<T> selector);

        IConsumer<T>
        SetExceptionHandler(IExceptionHandler handler);

        void
        Consume(T element);

        ISelector<T>
        GetSelector();

        IExceptionHandler
        GetExceptionHandler();

        int
        GetWaitingCount();
    }
}

//  ##########################################################################
